renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -0,0 +1,352 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "vk_acceleration_structure.h"
+#include "core/settings.h"
+#include "vk_core.h"
+
+RDOC_EXTERN_CONFIG(bool, Vulkan_Debug_SingleSubmitFlushing);
+
+namespace
+{
+// Although the serialised data is implementation-defined in general, the header is defined:
+// https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap37.html#vkCmdCopyAccelerationStructureToMemoryKHR
+constexpr std::size_t handleCountOffset = VK_UUID_SIZE + VK_UUID_SIZE + 8 + 8;
+constexpr VkDeviceSize handleCountSize = 8;
+
+// Spec says VkCopyAccelerationStructureToMemoryInfoKHR::dst::deviceAddress must be 256 bytes aligned
+constexpr VkDeviceSize asBufferAlignment = 256;
+}
+
+bool VulkanAccelerationStructureManager::Prepare(VkAccelerationStructureKHR as,
+                                                 const rdcarray<uint32_t> &queueFamilyIndices,
+                                                 ASMemory &result)
+{
+  const VkDeviceSize serialisedSize = SerialisedASSize(as);
+
+  const VkDevice d = m_pDriver->GetDev();
+  VkResult vkr = VK_SUCCESS;
+
+  // since this happens during capture, we don't want to start serialising extra buffer creates,
+  // leave this buffer as unwrapped
+  VkBuffer dstBuf = VK_NULL_HANDLE;
+
+  VkBufferCreateInfo bufInfo = {
+      VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+      NULL,
+      0,
+      serialisedSize,
+      VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT |
+          VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT,
+  };
+
+  // we make the buffer concurrently accessible by all queue families to not invalidate the
+  // contents of the memory we're reading back from.
+  bufInfo.sharingMode = VK_SHARING_MODE_CONCURRENT;
+  bufInfo.queueFamilyIndexCount = (uint32_t)queueFamilyIndices.size();
+  bufInfo.pQueueFamilyIndices = queueFamilyIndices.data();
+
+  // spec requires that CONCURRENT must specify more than one queue family. If there is only one
+  // queue family, we can safely use exclusive.
+  if(bufInfo.queueFamilyIndexCount == 1)
+    bufInfo.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+
+  vkr = ObjDisp(d)->CreateBuffer(Unwrap(d), &bufInfo, NULL, &dstBuf);
+  m_pDriver->CheckVkResult(vkr);
+
+  m_pDriver->AddPendingObjectCleanup(
+      [d, dstBuf]() { ObjDisp(d)->DestroyBuffer(Unwrap(d), dstBuf, NULL); });
+
+  VkMemoryRequirements mrq = {};
+  ObjDisp(d)->GetBufferMemoryRequirements(Unwrap(d), dstBuf, &mrq);
+
+  mrq.alignment = asBufferAlignment;
```

Comment:
I doubt buffer alignment would ever be worse than 256 bytes but this should be `RDCMAX(mrq.alignment, asBufferAlignment)` just to be safe.